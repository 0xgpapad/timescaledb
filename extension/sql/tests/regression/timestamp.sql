\set ON_ERROR_STOP 1

\o /dev/null
\ir include/create_clustered_db.sql

\o 
\set ECHO ALL
\c meta
SELECT add_cluster_user('postgres', NULL);

SELECT set_meta('meta' :: NAME, 'localhost');
SELECT add_node('Test1' :: NAME, 'localhost');
SELECT add_node('test2' :: NAME, 'localhost');

\c Test1

CREATE TABLE PUBLIC."testNs" (
  "timeCustom" TIMESTAMP NOT NULL,
  device_id TEXT NOT NULL,
  series_0 DOUBLE PRECISION NULL,
  series_1 DOUBLE PRECISION NULL,
  series_2 DOUBLE PRECISION NULL,
  series_bool BOOLEAN NULL
);
CREATE INDEX ON PUBLIC."testNs" (device_id, "timeCustom" DESC NULLS LAST) WHERE device_id IS NOT NULL;
SELECT * FROM create_hypertable('"public"."testNs"', 'timeCustom', 'device_id', hypertable_name=>'testNs', associated_schema_name=>'testNs' );

SELECT set_is_distinct_flag('"public"."testNs"', 'device_id', TRUE);


\c Test1
INSERT INTO "testNs"("timeCustom", device_id, series_0, series_1) VALUES
('2009-11-12T01:00:00+00:00', 'dev1', 1.5, 1),
('2009-11-12T01:00:00+00:00', 'dev1', 1.5, 2),
('2009-11-10T23:00:02+00:00', 'dev1', 2.5, 3);

SELECT close_chunk_end(c.id)
FROM get_open_partition_for_key('testNs', 'dev1') part
INNER JOIN chunk c ON (c.partition_id = part.id);

INSERT INTO "testNs"("timeCustom", device_id, series_0, series_1) VALUES
('2009-11-10T23:00:00+00:00', 'dev2', 1.5, 1),
('2009-11-10T23:00:00+00:00', 'dev2', 1.5, 2);

SELECT * FROM PUBLIC."testNs";


SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs'));

\echo 'The next 2 queries will differ in output between UTC and EST since the mod is on the 100th hour UTC'
SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((100 * 60 * 60 * 1e6) :: BIGINT)
                     ));
SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((100 * 60 * 60 * 1e6) :: BIGINT)
                     ));

\echo 'The rest of the queries will be the same in output between UTC and EST'

SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));
SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));



SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));
SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));


SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    time_condition=>new_time_condition(1257894000000000,1257987600000000) --microseconds
                  ));
SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    time_condition=>new_time_condition(1257894000000000,1257987600000000) --microseconds
                  ));


SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT),
                                    limit_time_periods => 2 
                  ));
SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(namespace_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT),
                                    limit_time_periods => 2
                  ));
