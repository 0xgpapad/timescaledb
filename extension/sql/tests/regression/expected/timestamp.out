\c meta
SELECT add_cluster_user('postgres', NULL);
 add_cluster_user 
------------------
 
(1 row)

SELECT set_meta('meta' :: NAME, 'localhost');
 set_meta 
----------
 
(1 row)

SELECT add_node('Test1' :: NAME, 'localhost');
 add_node 
----------
 
(1 row)

SELECT add_node('test2' :: NAME, 'localhost');
 add_node 
----------
 
(1 row)

\c Test1
CREATE TABLE PUBLIC."testNs" (
  "timeCustom" TIMESTAMP NOT NULL,
  device_id TEXT NOT NULL,
  series_0 DOUBLE PRECISION NULL,
  series_1 DOUBLE PRECISION NULL,
  series_2 DOUBLE PRECISION NULL,
  series_bool BOOLEAN NULL
);
CREATE INDEX ON PUBLIC."testNs" (device_id, "timeCustom" DESC NULLS LAST) WHERE device_id IS NOT NULL;
SELECT * FROM create_hypertable('"public"."testNs"', 'timeCustom', 'device_id', hypertable_name=>'testNs', associated_schema_name=>'testNs' );
  name  | main_schema_name | main_table_name | associated_schema_name | associated_table_prefix | root_schema_name | root_table_name | distinct_schema_name | distinct_table_name | replication_factor | placement | time_field_name |       time_field_type       | created_on | chunk_size_bytes 
--------+------------------+-----------------+------------------------+-------------------------+------------------+-----------------+----------------------+---------------------+--------------------+-----------+-----------------+-----------------------------+------------+------------------
 testNs | public           | testNs          | testNs                 | _hyper_1                | testNs           | _hyper_1_root   | testNs               | _hyper_1_distinct   |                  1 | STICKY    | timeCustom      | timestamp without time zone | Test1      |       1073741824
(1 row)

SELECT set_is_distinct_flag('"public"."testNs"', 'device_id', TRUE);
 set_is_distinct_flag 
----------------------
 
(1 row)

\c Test1
INSERT INTO "testNs"("timeCustom", device_id, series_0, series_1) VALUES
('2009-11-12T01:00:00+00:00', 'dev1', 1.5, 1),
('2009-11-12T01:00:00+00:00', 'dev1', 1.5, 2),
('2009-11-10T23:00:02+00:00', 'dev1', 2.5, 3);
SELECT _iobeamdb_meta_api.close_chunk_end_immediate(c.id)
FROM get_open_partition_for_key('testNs', 'dev1') part
INNER JOIN chunk c ON (c.partition_id = part.id);
 close_chunk_end_immediate 
---------------------------
 
(1 row)

INSERT INTO "testNs"("timeCustom", device_id, series_0, series_1) VALUES
('2009-11-10T23:00:00+00:00', 'dev2', 1.5, 1),
('2009-11-10T23:00:00+00:00', 'dev2', 1.5, 2);
SELECT * FROM PUBLIC."testNs";
     timeCustom      | device_id | series_0 | series_1 | series_2 | series_bool 
---------------------+-----------+----------+----------+----------+-------------
 2009-11-12 01:00:00 | dev1      |      1.5 |        1 |          | 
 2009-11-12 01:00:00 | dev1      |      1.5 |        2 |          | 
 2009-11-10 23:00:02 | dev1      |      2.5 |        3 |          | 
 2009-11-10 23:00:00 | dev2      |      1.5 |        1 |          | 
 2009-11-10 23:00:00 | dev2      |      1.5 |        2 |          | 
(5 rows)

SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs'));
                                                          json                                                          
------------------------------------------------------------------------------------------------------------------------
 {"timeCustom":"2009-11-12T01:00:00","device_id":"dev1","series_0":1.5,"series_1":1,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-12T01:00:00","device_id":"dev1","series_0":1.5,"series_1":2,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:02","device_id":"dev1","series_0":2.5,"series_1":3,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:00","device_id":"dev2","series_0":1.5,"series_1":1,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:00","device_id":"dev2","series_0":1.5,"series_1":2,"series_2":null,"series_bool":null}
(5 rows)

\echo 'The next 2 queries will differ in output between UTC and EST since the mod is on the 100th hour UTC'
The next 2 queries will differ in output between UTC and EST since the mod is on the 100th hour UTC
SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((100 * 60 * 60 * 1e6) :: BIGINT)
                     ));
                        json                        
----------------------------------------------------
 {"time":"2009-11-10T08:00:00","sum(series_0)":8.5}
(1 row)

SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((100 * 60 * 60 * 1e6) :: BIGINT)
                     ));
                        json                        
----------------------------------------------------
 {"time":"2009-11-10T03:00:00","sum(series_0)":8.5}
(1 row)

\echo 'The rest of the queries will be the same in output between UTC and EST'
The rest of the queries will be the same in output between UTC and EST
SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));
                        json                        
----------------------------------------------------
 {"time":"2009-11-10T23:00:00","sum(series_0)":5.5}
 {"time":"2009-11-12T01:00:00","sum(series_0)":3}
(2 rows)

SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));
                        json                        
----------------------------------------------------
 {"time":"2009-11-10T23:00:00","sum(series_0)":5.5}
 {"time":"2009-11-12T01:00:00","sum(series_0)":3}
(2 rows)

SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));
                        json                        
----------------------------------------------------
 {"time":"2009-11-10T23:00:00","sum(series_0)":5.5}
 {"time":"2009-11-12T01:00:00","sum(series_0)":3}
(2 rows)

SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT)
                     ));
                        json                        
----------------------------------------------------
 {"time":"2009-11-10T23:00:00","sum(series_0)":5.5}
 {"time":"2009-11-12T01:00:00","sum(series_0)":3}
(2 rows)

SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    time_condition=>new_time_condition(1257894000000000,1257987600000000) --microseconds
                  ));
                                                          json                                                          
------------------------------------------------------------------------------------------------------------------------
 {"timeCustom":"2009-11-10T23:00:02","device_id":"dev1","series_0":2.5,"series_1":3,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:00","device_id":"dev2","series_0":1.5,"series_1":1,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:00","device_id":"dev2","series_0":1.5,"series_1":2,"series_2":null,"series_bool":null}
(3 rows)

SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    time_condition=>new_time_condition(1257894000000000,1257987600000000) --microseconds
                  ));
                                                          json                                                          
------------------------------------------------------------------------------------------------------------------------
 {"timeCustom":"2009-11-10T23:00:02","device_id":"dev1","series_0":2.5,"series_1":3,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:00","device_id":"dev2","series_0":1.5,"series_1":1,"series_2":null,"series_bool":null}
 {"timeCustom":"2009-11-10T23:00:00","device_id":"dev2","series_0":1.5,"series_1":2,"series_2":null,"series_bool":null}
(3 rows)

SET timezone = 'UTC';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT),
                                    limit_time_periods => 2
                  ));
                       json                       
--------------------------------------------------
 {"time":"2009-11-12T01:00:00","sum(series_0)":3}
(1 row)

SET timezone = 'EST';
SELECT *
FROM ioql_exec_query(new_ioql_query(hypertable_name => 'testNs',
                                    select_items => ARRAY [new_select_item('series_0', 'SUM')],
                                    aggregate => new_aggregate((1 * 60 * 60 * 1e6) :: BIGINT),
                                    limit_time_periods => 2
                  ));
                       json                       
--------------------------------------------------
 {"time":"2009-11-12T01:00:00","sum(series_0)":3}
(1 row)

